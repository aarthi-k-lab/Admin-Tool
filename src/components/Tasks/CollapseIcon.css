.collapse-icon-container {
  align-items: center;
  cursor: pointer;
  display: flex;
  position: relative;
  white-space: nowrap;
  width: 0.8rem;
  float: right;
  padding-right: 0.25rem;
  padding-top: 0.25rem;
}

.shift-left {
  position: absolute;
  left: 0.2rem;
}
