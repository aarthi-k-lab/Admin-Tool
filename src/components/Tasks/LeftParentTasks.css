
.left-parent-tasks {
  width: 100%;
  transition: width 250ms ease-in;
  animation: fadein 0.5s;
}

.status-select-field-span {
  display: inline-flex;
  margin-left: 1.5rem;
}

.status-select-field-text {
  font-size: 4px !important;
}

.status-text {
  margin: 0.4rem;
}

.checked-circle :checked {
  color: green !important;
}

.subtask-list {
  margin-left: 2rem;
}

.parent-task-name {
  display: inline-block;
  font-size: 16px;
  font-weight: 600;
  vertical-align: middle;
  margin-top: 1rem;
}

.parent-task-span {
  display: inline-block;
}

.subtask-name {
  display: inline-block;
  margin-top: 1rem;
  font-size: 0.9rem;
}

.iconarea {
  align-self: center;
}

.left-pane-tasks-main-div {
  margin-left: 0.5rem;
  overflow-y: auto;
  overflow-x: hidden;
  align-self: flex-start;
  display: inline-flex;
  flex-direction: column;
  height: 100%;
  min-height: calc(100vh - 15rem);
  max-height: calc(100vh - 14.9rem);
}

.subtask-grid {
  margin: -0.75rem 0 -1rem -0.8rem;
}

.parent-task-grid-div {
  margin-bottom: -0.5rem;
}

@-webkit-keyframes fadein {
  from { opacity: 0.5; }
  to { opacity: 1; }
}

.error-text-subtask {
  font-size: 12px;
  font-weight: 200;
  color: var(--red-a700);
}

.error-text-parent {
  font-size: 14px;
  font-weight: 200;
  color: var(--red-a700);
}

.failure-subtask-name {
  display: inline-block;
  padding-top: 0.5rem;
  color: var(--red-a700);
  font-size: 0.9rem;
}
