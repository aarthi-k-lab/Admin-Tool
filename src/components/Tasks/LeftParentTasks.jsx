/* eslint-disable no-underscore-dangle */
import React from 'react';
import Grid from '@material-ui/core/Grid';
import PropTypes from 'prop-types';
import * as R from 'ramda';
import TaskModel from 'lib/PropertyValidation/TaskModel';
import hotkeys from 'hotkeys-js';
import TaskStatusIcon from '../TaskStatusIcon';
import SubTask from './SubTask/SubTask';
import DeleteTask from './OptionalTask/DeleteTask';
import styles from './LeftParentTasks.css';

const SHIFT_DOWN_KEY = [16, 40];
const SHIFT_UP_KEY = [16, 38];
class LeftParentTasks extends React.Component {
  constructor(props) {
    super(props);
    this.renderTasks = this.renderTasks.bind(this);
    const { optionalTasks } = this.props;
    this.state = {
      isTaskAdded: optionalTasks.map(task => task.visibility),
    };
    this.changedTask = {
      taskIdx: 0,
      task: {},
    };
  }

  componentDidMount() {
    hotkeys('*', (event) => {
      if (event.type === 'keydown') {
        this.handleHotKeyPress();
      }
    });
  }

  componentDidUpdate(prevProps) {
    const { shouldDeleteTask, resetDeleteTaskConfirmation } = this.props;
    const { taskIdx, task } = this.changedTask;
    if (prevProps.shouldDeleteTask !== shouldDeleteTask) {
      if (shouldDeleteTask) {
        this.modifyTaskList(taskIdx, task, 'DELETE');
        resetDeleteTaskConfirmation();
      }
    }
  }

  componentWillUnmount() {
    hotkeys.unbind('*');
  }

  findLastItem= (tasks) => {
    let lastItem = 0;
    tasks.forEach((task, index) => {
      lastItem = task.visibility ? index : lastItem;
    });
    return lastItem;
  }

  handleHotKeyPress = () => {
    const { tasks, onSubTaskClick, selectedTaskId } = this.props;
    if (R.equals(hotkeys.getPressedKeyCodes(), SHIFT_DOWN_KEY)) {
      const subtaskOrder = this.findLastItem(tasks);
      const lastItem = this.findLastItem(tasks[subtaskOrder].subTasks);
      if (!R.equals(selectedTaskId, tasks[subtaskOrder].subTasks[lastItem]._id)) {
        onSubTaskClick(tasks[subtaskOrder].subTasks[lastItem]._id);
      }
    } else if (R.equals(hotkeys.getPressedKeyCodes(), SHIFT_UP_KEY)) {
      if (!R.equals(selectedTaskId, tasks[0].subTasks[0]._id)) {
        onSubTaskClick(tasks[0].subTasks[0]._id);
      }
    }
  }

  deleteTask(taskIdx, task) {
    this.changedTask = Object.assign({}, this.changedTask, { taskIdx, task });
    const { handleShowDeleteTaskConfirmation } = this.props;
    const payload = {
      deleteTaskConfirmationDialog: {
        title: 'DELETE TASK',
        isOpen: true,
        content: 'Deleting a task will delete all the associated checklist information. Do you like to proceed?',
      },
    };
    handleShowDeleteTaskConfirmation(payload);
  }

  modifyTaskList(taskIdx, task, type) {
    const { updateChecklist } = this.props;
    const { isTaskAdded } = this.state;
    const isTaskAddedList = isTaskAdded;
    isTaskAddedList[taskIdx] = !isTaskAddedList[taskIdx];
    this.setState({
      isTaskAdded: isTaskAddedList,
    });
    const payload = {
      task: Object.assign({}, task, { visibility: isTaskAdded[taskIdx] }),
      fieldName: 'visibility',
      type,
    };
    updateChecklist(payload);
  }

  renderDeleteIcon(task) {
    const { disableModifyOptionalTasks, optionalTasks } = this.props;
    // eslint-disable-next-line no-underscore-dangle
    const index = optionalTasks.findIndex(optTask => optTask.id === task._id);
    if (index !== -1) {
      return (
        <DeleteTask
          disabled={disableModifyOptionalTasks}
          margin={{ 'margin-right': '2.2rem' }}
          onClick={() => this.deleteTask(index, task)}
          toolTipPosition="left"
        />
      );
    }
    return null;
  }

  static renderCollapsedView(task) {
    return (
      <Grid container spacing={0}>
        <Grid item xs={12}>
          <span placement="right" title={R.pathOr('', ['taskBlueprint', 'name'], task)}>
            <TaskStatusIcon styleName="icon-padding" task={task} />
          </span>
        </Grid>
      </Grid>
    );
  }

  renderTasksChecklist(task, selectedTaskId, onSubTaskClick) {
    const isTaskSelected = task.subTasks.some(({ _id: id }) => id === selectedTaskId);
    return (
      <>
        <Grid
          className={
            isTaskSelected
              ? styles['selected-task']
              : ''
          }
          container
          spacing={0}
          wrap="nowrap"
        >
          <Grid alignItems="center" container item justify="center" xs={2}>
            <TaskStatusIcon isSelected={isTaskSelected} styleName="fill-width" task={task} />
          </Grid>
          <Grid alignItems="center" container item xs={10}>
            <span styleName="parent-task-name">{ R.pathOr('', ['taskBlueprint', 'name'], task) }</span>
            {
              task.failureReason
                && (
                  <>
                    <br />
                    <span styleName="error-text-parent">{ task.failureReason }</span>
                  </>
                )
            }
          </Grid>
          <Grid alignItems="center" container item xs={2}>
            {this.renderDeleteIcon(task)}
          </Grid>
        </Grid>
        {
          task.subTasks && task.subTasks.length ? (
            <Grid container direction="column" spacing={0}>
              {
                task.subTasks
                  .filter(({ visibility }) => visibility)
                  .map(subTask => (
                    <SubTask
                      key={subTask._id} // eslint-disable-line
                      data={subTask}
                      onClick={onSubTaskClick}
                      selected={subTask._id === selectedTaskId} // eslint-disable-line
                    />
                  ))
              }
            </Grid>
          ) : null
        }
      </>
    );
  }

  renderTasks(isCollapsed) {
    const {
      tasks, onSubTaskClick, selectedTaskId,
    } = this.props;
    return (
      tasks
        .filter(({ visibility }) => visibility)
        .map(task => (
          <div key={task._id} styleName="task-group">
            {
            isCollapsed
              ? this.constructor.renderCollapsedView(task)
              : this.renderTasksChecklist(task,
                selectedTaskId, onSubTaskClick)
            }
          </div>
        ))
    );
  }

  render() {
    const { isCollapsed } = this.props;
    return (
      <div styleName="left-pane-tasks-main-div">
        <div styleName="left-pane-scrollable-tasks">
          {
            this.renderTasks(isCollapsed)
          }
        </div>
      </div>
    );
  }
}

LeftParentTasks.defaultProps = {
  disableModifyOptionalTasks: false,
  selectedTaskId: '',
};

LeftParentTasks.propTypes = {
  disableModifyOptionalTasks: PropTypes.bool,
  handleShowDeleteTaskConfirmation: PropTypes.func.isRequired,
  isCollapsed: PropTypes.bool.isRequired,
  onSubTaskClick: PropTypes.func.isRequired,
  optionalTasks: PropTypes.arrayOf(PropTypes.shape).isRequired,
  resetDeleteTaskConfirmation: PropTypes.func.isRequired,
  selectedTaskId: PropTypes.string,
  shouldDeleteTask: PropTypes.bool.isRequired,
  tasks: PropTypes.arrayOf(TaskModel).isRequired,
  updateChecklist: PropTypes.func.isRequired,
};


export default LeftParentTasks;
