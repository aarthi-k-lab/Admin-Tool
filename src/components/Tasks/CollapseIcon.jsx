import React from 'react';
import ChervronLeft from '@material-ui/icons/ChevronLeft';

import './CollapseIcon.css';

const CollapseIcon = () => (
  <span styleName="collapse-icon-container">
    <ChervronLeft />
    <ChervronLeft styleName="shift-left" />
  </span>
);

export default CollapseIcon;
