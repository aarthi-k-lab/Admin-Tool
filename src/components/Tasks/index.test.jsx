import React from 'react';
import renderer from 'react-test-renderer';
import CollapseIcon from './CollapseIcon';
import LeftParentTasks from './LeftParentTasks';

describe('<Task Components />', () => {
  it('CollapseIcon renders correctly', () => {
    const tree = renderer
      .create(<CollapseIcon direction="left" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('LeftParentTasks renders correctly when expanded', () => {
    const tree = renderer
      .create(<LeftParentTasks isCollapsed={false} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('LeftParentTasks renders correctly when collapsed', () => {
    const tree = renderer
      .create(<LeftParentTasks isCollapsed />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
