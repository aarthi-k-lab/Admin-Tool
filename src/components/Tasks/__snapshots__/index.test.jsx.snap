// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Task Components /> CollapseIcon renders correctly 1`] = `
<svg
  aria-hidden="true"
  className="MuiSvgIcon-root-1"
  focusable="false"
  role="presentation"
  viewBox="0 0 24 24"
>
  <path
    d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;

exports[`<Task Components /> LeftParentTasks renders correctly when collapsed 1`] = `
<div
  className="components-Tasks-LeftParentTasks__left-pane-tasks-main-div"
>
  <div
    className="components-Tasks-LeftParentTasks__left-pane-scrollable-tasks"
  />
</div>
`;

exports[`<Task Components /> LeftParentTasks renders correctly when expanded 1`] = `
<div
  className="components-Tasks-LeftParentTasks__left-pane-tasks-main-div"
>
  <div
    className="components-Tasks-LeftParentTasks__left-pane-scrollable-tasks"
  />
</div>
`;
